orchagent/nhgorch.h

```diff
@@ -1,249 +1,127 @@
 
     /* Check if the next hop is labeled. */
-    inline bool isLabeled() const { return !m_nh_key.label_stack.empty(); }
+    inline bool isLabeled() const { return !m_key.label_stack.empty(); }
 
     /* Convert member's details to string. */
-    std::string to_string() const
+    string to_string() const override
     {
-        return m_nh_key.to_string() + ", SAI ID: " + std::to_string(m_gm_id);
+        return m_key.to_string() + ", SAI ID: " + std::to_string(m_gm_id);
     }
-
-private:
-    /* The key of the next hop of this member. */
-    NextHopKey m_nh_key;
-
-    /* The group member SAI ID for this member. */
-    sai_object_id_t m_gm_id;
 };
 
-/* Map indexed by NextHopKey, containing the SAI ID of the group member. */
-typedef std::map<NextHopKey, NextHopGroupMember> NhgMembers;
-
 /*
  * NextHopGroup class representing a next hop group object.
  */
-class NextHopGroup
+class NextHopGroup : public NhgCommon<NextHopGroupKey, NextHopKey, NextHopGroupMember>
 {
 public:
     /* Constructors. */
     explicit NextHopGroup(const NextHopGroupKey& key, bool is_temp);
-    NextHopGroup(NextHopGroup&& nhg);
+
+    NextHopGroup(NextHopGroup&& nhg) :
+        NhgCommon(move(nhg)), m_is_temp(nhg.m_is_temp)
+    { SWSS_LOG_ENTER(); }
+
     NextHopGroup& operator=(NextHopGroup&& nhg);
 
     /* Destructor. */
     virtual ~NextHopGroup() { remove(); }
 
     /* Sync the group, creating the group's and members SAI IDs. */
-    bool sync();
+    bool sync() override;
 
     /* Remove the group, reseting the group's and members SAI IDs.  */
-    bool remove();
+    bool remove() override;
 
     /*
      * Update the group based on a new next hop group key.  This will also
      * perform any sync / remove necessary.
      */
     bool update(const NextHopGroupKey& nhg_key);
 
-    /* Check if the group contains the given next hop. */
-    inline bool hasNextHop(const NextHopKey& nh_key) const
-    {
-        return m_members.find(nh_key) != m_members.end();
-    }
-
     /* Validate a next hop in the group, syncing it. */
     bool validateNextHop(const NextHopKey& nh_key);
 
     /* Invalidate a next hop in the group, removing it. */
     bool invalidateNextHop(const NextHopKey& nh_key);
 
-    /* Increment the number of existing groups. */
-    static inline void incCount() { ++m_count; }
-
-    /* Decrement the number of existing groups. */
-    static inline void decCount() { assert(m_count > 0); --m_count; }
-
     /* Getters / Setters. */
-    inline const NextHopGroupKey& getKey() const { return m_key; }
-    inline sai_object_id_t getId() const { return m_id; }
-    static inline unsigned int getCount() { return m_count; }
-    inline bool isTemp() const { return m_is_temp; }
-    inline bool isSynced() const { return m_id != SAI_NULL_OBJECT_ID; }
-    inline size_t getSize() const { return m_members.size(); }
+    inline bool isTemp() const override { return m_is_temp; }
```

Comment:
In what scenario NHG is created as temp?

Reply:
If there isn't space in the data plane to create another next hop group, we pick a next hop within the group and use its ID as the group's ID. When space becomes available we try and create the actual next hop group. This is the same mechanism as routeOrch.